lib/bap_disasm/bap_disasm_rec.ml

```diff
@@ -336,72 +310,88 @@ let add_destinations bil = function
-      | None when Addrs.mem leads (next finish) &&
-                  not (Addrs.mem stage1.dests finish) ->
-        Addrs.add_multi preds ~key:(next finish) ~data:start;
-        [Some (next finish),`Fall]
+      | None when Addrs.mem leads (next max_addr) &&
+                  not (Addrs.mem stage1.dests addr) ->
+        Addrs.add_multi preds ~key:(next max_addr) ~data:start;
+        [Some (next max_addr),`Fall]
       | None -> [] in
     Addrs.add_exn succs ~key:start ~data:dests;
     return () in
-
-  let rec loop start curr' =
-    let curr = next curr' in
-    if is_visited curr then
-      if is_edge curr then
-        create_block start curr >>= fun () ->
-        loop (next curr) curr
-      else loop start curr
-    else match next_visited curr with
-      | Some addr -> loop addr addr
-      | None when is_visited start && is_visited curr' ->
-        create_block start curr'
-      | None -> return () in
-  match Span.min stage1.visited with
-  | None -> errorf "Provided memory doesn't contain a recognizable code"
-  | Some addr -> loop addr addr >>= fun () ->
-    let dis = Dis.store_asm dis in
-    let dis = Dis.store_kinds dis in
-    let disasm mem =
-      Dis.run dis mem
-        ~init:[] ~return:ident ~stopped:(fun s _ ->
-            Dis.stop s (Dis.insns s)) |>
-      List.map ~f:(function
-          | mem, None -> mem,(None,None)
-          | mem, (Some ins as insn) ->
-            let dests =
-              Addrs.find stage1.dests (Memory.max_addr mem) |>
-              Option.value ~default:[] |>
-              List.filter_map ~f:(function
-                  | a, (`Cond | `Jump) -> a
-                  | _ -> None) in
-            let bil = match stage1.lift mem ins with
-              | Ok bil -> bil
-              | _ -> [] in
-            mem, (insn, Some (add_destinations bil dests))) in
-    return {stage1; addrs; succs; preds; disasm}
+  let next_block leftovers start =
+    let rec loop leftovers last start curr =
+      match Visited.find_insn leftovers curr with
+      | Some max_addr ->
+        let insn = curr,max_addr in
+        let leftovers = Visited.remove_insn leftovers curr in
+        if is_edge curr max_addr then Some (leftovers, start, insn)
+        else loop leftovers (Some insn) start (next max_addr)
+      | None -> match last with
+        | Some last when is_insn start -> Some (leftovers,start,last)
+        | _ -> match next_visited curr with
+          | Some (addr,_) -> loop leftovers None addr addr
+          | None -> None in
+    loop leftovers None start start in
+  let fetch_blocks () =
+    let rec loop leftovers start =
+      match next_block leftovers start with
+      | Some (leftovers,start,(addr, max_addr)) ->
+         if Hashtbl.mem addrs start then
+           loop leftovers (next max_addr)
+         else
+           create_block start addr max_addr >>= fun () ->
+           loop leftovers (next max_addr)
+      | None when Map.is_empty leftovers -> Ok ()
```

Comment:
it looks like that this case is not necessary, it will be handled by the case on line 389